### PR TITLE
Implement cart page

### PR DESCRIPTION
**Pull request by Alesia-V175**

### CODE REVIEW

File: src/components/main/cart/Cart.scss [hunk 1]
```diff
@@ -67,10 +67,12 @@
   }
 
   .product-card {
+    width: 100%;
     display: flex;
     justify-content: space-around;
     border-bottom: 1px solid #CACDD8;
     padding: 15px 0;
+    box-sizing: border-box;
```
Comment (BondPV) [inline src/components/main/cart/Cart.scss]: property ( box-sizing: border-box;) added globally in _main.scss

File: src/components/main/grid/Grid.ts [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import './grid.scss';
 import { ClassListName, ClassMap } from 'constants/htmlConstants';
+import LocalStorage from 'helpers/localStorage/LocalStorage';
 
 class Grid {
   public changeGridList(): void {
+    LocalStorage.saveGridCatalog('list');
```
Comment (BondPV) [inline src/components/main/grid/Grid.ts]: Not relevant for the current branch 'develop'

File: src/components/main/grid/Grid.ts [hunk 3]
```diff
@@ -45,6 +48,14 @@ class Grid {
     viewButtonList.classList.add(ClassListName.catalogView, ClassListName.catalogViewList);
     catalogHeader.append(viewButtonList);
 
+    const currentGridValue = LocalStorage.getGridCatalog();
```
Comment (BondPV) [inline src/components/main/grid/Grid.ts]: Not relevant for the current branch 'develop'

File: src/components/main/sortCatalog/SortCatalog.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import './sortCatalog.scss';
 import { IProduct } from 'types/interfaces';
 import { SortOption } from 'types/types';
 import { SortOptionMap, optionNames, ClassMap, valueOptionAsc, ClassListName } from 'constants/htmlConstants';
+import LocalStorage from 'helpers/localStorage/LocalStorage';
```
Comment (BondPV) [inline src/components/main/sortCatalog/SortCatalog.ts]: Not relevant for the current branch 'develop'

File: src/components/main/cart/Cart.scss [hunk 5]
```diff
@@ -0,0 +1,28 @@
+.cart-wrap {
```
Comment (BondPV) [inline src/components/main/cart/Cart.scss]: Class names don't match BEM

**Review by BondPV — APPROVED** [review]

File: src/components/main/cart/Cart.scss (hunk 1 above)
Comment (Alesia-V175) [reply src/components/main/cart/Cart.scss]: deleted unnecessary property

File: src/components/main/grid/Grid.ts (hunk 2 above)
Comment (Alesia-V175) [reply src/components/main/grid/Grid.ts]: fixed

File: src/components/main/grid/Grid.ts (hunk 3 above)
Comment (Alesia-V175) [reply src/components/main/grid/Grid.ts]: fixed

File: src/components/main/sortCatalog/SortCatalog.ts (hunk 4 above)
Comment (Alesia-V175) [reply src/components/main/sortCatalog/SortCatalog.ts]: fixed

File: src/components/main/cart/Cart.scss (hunk 5 above)
Comment (Alesia-V175) [reply src/components/main/cart/Cart.scss]: fixed

File: src/types/enums.ts [hunk 6]
```diff
@@ -22,3 +22,24 @@ export enum Symbol {
   Query = '?',
   Equality = '=',
 }
+
+export enum CartText {
+  ShoppingCart = 'Shopping Cart',
+  ShowPage = 'Show per page ',
+  Summary = 'Summary',
+
```
Comment (Ksarelto) [inline src/types/enums.ts:30]: for what is this empty line?

File: src/components/main/catalog/Catalog.ts [hunk 7]
```diff
@@ -8,13 +8,17 @@ class Catalog {
 
   render() {
     const parentElement = document.querySelector(IdMap.catalog);
-    const products: string = this.products
+    const products: HTMLElement[] = this.products
       .map((product) => new Product(product))
-      .map((product) => product.render())
-      .join('');
+      .map((product) => product.render());
```
Comment (Ksarelto) [inline src/components/main/catalog/Catalog.ts:13]: y can call 'render' method in previous map

File: src/components/main/catalog/Catalog.ts [hunk 8]
```diff
@@ -8,13 +8,17 @@ class Catalog {
...
 
     if (parentElement instanceof HTMLElement) {
```
Comment (Ksarelto) [inline src/components/main/catalog/Catalog.ts:15]: no nested 'if's'

File: src/components/main/cart/cart.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (cartTotalPrice && cartTotalQty) {
+      cartTotalQty.textContent = `Total products: ${totalQty}`;
+      cartTotalPrice.textContent = `Total: $ ${totalPrice}`;
+    }
+
+    if (cartDiscountPrice && cartPriceWrap && cartTotalPrice) {
+      this.createDiscountItems();
+
+      if (LocalStorage.getPromo().length > 0) {
```
Comment (Ksarelto) [inline src/components/main/cart/cart.ts:34]: try to not create nested ifs

File: src/components/main/productDetails/ProductDetails.ts [hunk 10]
```diff
@@ -40,16 +41,29 @@ class ProductDetails extends Product {
 
     const buttonAddToCart: HTMLElement = document.createElement('button');
     buttonAddToCart.classList.add('product-details__button');
-    buttonAddToCart.innerText = 'ADD TO CARD';
+    buttonAddToCart.textContent = 'ADD TO CART';
```
Comment (Ksarelto) [inline src/components/main/productDetails/ProductDetails.ts:44]: this is a magic string

File: src/components/main/productDetails/ProductDetails.ts [hunk 11]
```diff
@@ -40,16 +41,29 @@ class ProductDetails extends Product {
...
     buttons.append(buttonAddToCart);
 
     const buttonBuyNow: HTMLElement = document.createElement('button');
     buttonBuyNow.classList.add('product-details__button', 'product-details__button_buy');
     buttonBuyNow.innerText = 'BUY NOW';
```
Comment (Ksarelto) [inline src/components/main/productDetails/ProductDetails.ts:49]: the same as above

File: src/components/main/productDetails/ProductDetails.ts [hunk 12]
```diff
@@ -40,16 +41,29 @@ class ProductDetails extends Product {
...
+      buttonAddToCart.classList.remove('product-details__button_remove');
+      buttonAddToCart.textContent = 'ADD TO CART';
+    }
+
     buttonAddToCart.addEventListener('click', () => {
-      alert('click Add to cart');
+      if (buttonAddToCart.classList.length === 1) {
+        buttonAddToCart.textContent = 'REMOVE FROM CART';
+      } else {
+        buttonAddToCart.textContent = 'ADD TO CART';
```
Comment (Ksarelto) [inline src/components/main/productDetails/ProductDetails.ts:64]: this button names should be in enum or object

File: src/helpers/localStorage/LocalStorage.ts [hunk 13]
```diff
@@ -1,3 +1,67 @@
...
+
+  public static removeProductFromCart(id: number): void {
+    const cartArr = this.getCart();
+    const filteredCart = cartArr.filter((product) => product.id !== id);
+    localStorage.setItem('cart', JSON.stringify(filteredCart));
+  }
+
+  public static getCart(): ICartProduct[] {
+    const cartString = localStorage.getItem('cart');
+    if (cartString) {
```
Comment (Ksarelto) [inline src/helpers/localStorage/LocalStorage.ts:29]: space around 'if'

File: src/helpers/localStorage/LocalStorage.ts [hunk 14]
```diff
@@ -1,3 +1,67 @@
...
+        item.count = count;
+      }
+    });
+    localStorage.setItem('cart', JSON.stringify(cartProducts));
+  }
+
+  public static removeProductFromCart(id: number): void {
+    const cartArr = this.getCart();
+    const filteredCart = cartArr.filter((product) => product.id !== id);
+    localStorage.setItem('cart', JSON.stringify(filteredCart));
```
Comment (Ksarelto) [inline src/helpers/localStorage/LocalStorage.ts:24]: if u use this name of key for localstorage in other places, better to create a constant

File: src/helpers/localStorage/LocalStorage.ts [hunk 15]
```diff
@@ -1,3 +1,67 @@
...
+    const cartString = localStorage.getItem('cart');
+    if (cartString) {
+      return JSON.parse(cartString);
+    }
+    return [];
+  }
+
+  public static isProductExists(id: number): boolean {
+    const cartArr = this.getCart();
+    return !!cartArr.find((product) => product.id === id);
```
Comment (Ksarelto) [inline src/helpers/localStorage/LocalStorage.ts:37]: u cat write like this
```
return !!this.getCart().find((product) => product.id === id);
```

File: src/helpers/localStorage/LocalStorage.ts [hunk 16]
```diff
@@ -1,3 +1,67 @@
...
+  }
+
+  public static getPromo(): string[] {
+    const promoStorage = localStorage.getItem('promo');
+    if (promoStorage) {
```
Comment (Ksarelto) [inline src/helpers/localStorage/LocalStorage.ts:42]: space around 'if'

File: src/helpers/localStorage/LocalStorage.ts [hunk 17]
```diff
@@ -1,3 +1,67 @@
...
+
+  public static removePromo(promo: string): void {
+    const promoStorage = this.getPromo();
+    const filteredPromo = promoStorage.filter((item) => item !== promo);
+    localStorage.setItem('promo', JSON.stringify(filteredPromo));
+  }
+
+  public static isPromoExist(promo: string): boolean {
+    const promoStorage = this.getPromo();
+    console.log(promoStorage);
```
Comment (Ksarelto) [inline src/helpers/localStorage/LocalStorage.ts:62]: no need in console.log

File: src/helpers/localStorage/LocalStorage.ts [hunk 18]
```diff
@@ -1,3 +1,67 @@
...
+  public static addPromo(promo: string): void {
+    const promoStorage = this.getPromo();
+    promoStorage.push(promo);
+    localStorage.setItem('promo', JSON.stringify(promoStorage));
+  }
+
+  public static removePromo(promo: string): void {
+    const promoStorage = this.getPromo();
+    const filteredPromo = promoStorage.filter((item) => item !== promo);
+    localStorage.setItem('promo', JSON.stringify(filteredPromo));
```
Comment (Ksarelto) [inline src/helpers/localStorage/LocalStorage.ts:57]: create a constant for name of the key fir localstorage

**Review by Ksarelto — COMMENTED** [review]
It is veeery big PR, it is very hard to check

**Pull request merged**